data/blog/js-mutation-testing-our-sdks.md

```diff
@@ -0,0 +1,166 @@
+Mutation testing (MT) is considered the gold standard when it comes to evaluating the fault detection ability (I'm going to stop with academic terms in a sec!) of your tests.
+In this article, we'll explore what mutation testing is, why it's a rather niche technique and what we learnt from using it on our JavaScript SDK codebase.
+
+## Intro to Mutation Testing
+
+You probably heard of _code coverage_ before, which comes in various flavours and granularities (statement, line, branch coverage, etc).
+Perhaps you use it in your codebase and you're happy about that green >90% coverage badge in your repo.
+While that is great, coverage has a substantial limitation: It only tells you which units of code (e.g. lines) were executed during a test run.
+It tells you nothing about if the executed lines were actually _checked_ or if—were you to introduce a bug into one of these lines—your tests would actually catch the bug.
+So playing devil's advocate, you could say coverage really only tells you one thing with certainty: Which parts of your code are _not_ covered by any test.
+
+Let's circle back to the part about coverage not telling you if an introduced bug would be caught. What if there was a tool that does exactly that? Enter mutation testing.
+
+The idea of mutation testing is fairly simple:
+
+1. Make a slight modification to your codebase. (In MT terms: Apply a _mutant operator_ to your code base.) For example, change an equality operator:
+   ```js
+   function isEven(num) {
+     // return num % 2 === 0; // mutate to:
+     return num % 2 !== 0
+   }
+   ```
+   The resulting code is called a "mutant".
+2. Run your tests against the mutant.
+   - If at least one of your test fails, crashes or times out, the mutant was _killed_.
+   - If all your tests still pass without a crash or timeout, the mutant _survived_.
+3. Repeat 1 and 2 until all possible mutant operators have been applied to your codebase or you reached a pre-defined maximum number of mutants.
+
+After this virtual bloodbath, you can calculate a _Mutation Score_ by dividing the number of killed mutants by the number of all created mutants. The score now tells you how likely it is that your tests would catch an actual bug.
+
+### Mutant Similarity Assumption
+
+The entire concept of MT depends on the assumption that [mutants are similar to actual bugs](https://dl.acm.org/doi/abs/10.1145/2635868.2635929) that programmers might introduce into the code base.
+Now the question is, is this true in reality? Luckily, engineers at Google publish papers about mutation testing every now and then. In [one of them](https://dl.acm.org/doi/10.1109/ICSE43902.2021.00087), they
+investigated the similarity assumption by comparing 15 million mutants against bugfix PRs. They found a high similarity of mutants to bugs, suggesting that the assumption does hold up.
+Does this translate to other code bases? Maybe not, but for the sake of trying things out, let's go with "yes".
```

Comment:
Maybe you can re-phrase this, instead of saying "maybe not, but let's assume yes" you could write something like:

Does this translate to other code bases? It's hard to say with certainty, but it seems reasonable to assume that there is at least some level of relevance to most code bases.